Sources/ApodiniNetworking/HTTPPathMatcher.swift

```diff
@@ -0,0 +1,171 @@
+//
+// This source file is part of the Apodini open source project
+//
+// SPDX-FileCopyrightText: 2019-2021 Paul Schmiedmayer and the Apodini project authors (see CONTRIBUTORS.md) <paul.schmiedmayer@tum.de>
+//
+// SPDX-License-Identifier: MIT
+//
+
+import ApodiniUtils
+
+
+struct HTTPPathMatcher {
+    struct MatchResult: Hashable {
+        /// The penalty associated with this match. Lower scores are better.
+        fileprivate(set) var penaltyScore: Int = 0
+        fileprivate(set) var parameters = HTTPRequest.ParametersStorage()
+    }
+    
+    
+    // Config
+    /// Whether or not to disable case sensitivity requirements when matching verbatim (i.e. non-parameter and non-wildcard) path components
+    private let allowsCaseInsensitiveMatching: Bool
+    /// Whether or not the matcher should allow multi-component wildcards (e.g.: `**`) to also match zero path components
+    private let allowsEmptyMultiWildcards: Bool
+    
+    // Input
+    private let urlComponents: [String]
+    private let pathComponents: [HTTPPathComponent]
+    
+    // State
+    private var urlComponentsIdx: Int = 0
+    private var pathComponentsIdx: Int = 0
+    
+    // Output
+    private var wipResult = MatchResult()
+    private var potentialMatches: Set<MatchResult> = []
+    
+    
+    static func match(
+        url: URI,
+        against inputPathComponents: [HTTPPathComponent],
+        allowsCaseInsensitiveMatching: Bool,
+        allowsEmptyMultiWildcards: Bool
+    ) -> HTTPRequest.ParametersStorage? {
+        var path = url.path
+        if path.hasPrefix("/") && path.count > 1 {
+            path.removeFirst()
+        }
+        if path.isEmpty || path == "/" {
+            // We're accessing the root path, which means that there are no parameters, be it named or wildcard.
+            if inputPathComponents.isEmpty {
+                return .init()
+            } else {
+                return nil
+            }
+        }
+        var matcher = Self(
+            allowsCaseInsensitiveMatching: allowsCaseInsensitiveMatching,
+            allowsEmptyMultiWildcards: allowsEmptyMultiWildcards,
+            urlComponents: path.components(separatedBy: "/"),
+            pathComponents: inputPathComponents
+        )
+        return matcher.match()?.parameters
+    }
+    
+    
+    private enum MatchOneResult {
+        case abort
+        case successAndContinue
+        case successAndReturn
+    }
+    
+    private mutating func match() -> MatchResult? {
```

Comment:
Out of interest, how does the matcher implementation compare to the Vapor one (I don't now that one). Just remember, that Vapor said about themselves that they got a pretty fast routing system. Would just be curious how those compare and how those two implementations compare performance wise.

Reply:
I haven't done any performance testing, but my gut response would be that it's probably worse.
There's some ways this probably could be improved (somehow caching past routing results comes to mind), but we're currently not doing any of that.

Reply:
@Supereg addendum: routing performance should depend primarily on whether or not you're using routes w/ multi-component wildcards. if you are not, it's at worst O(n) w.r.t. the number of routes registered for the request's HTTP method (and, in the case of HEAD requests, also the number of GET routes registered[^1]). there are, of course, early exits for cases where a URL obviously doesn't match a route, see the non-private function above, but at worst it'll have to compare all path components in the URL and the route. 

for routes with multi-component wildcards, performance is by definition significantly worse (compared to my non-wildcard matching, not compared to Vapor, although that probably as well). the reason being that there's multiple ways a URL could be matched onto a multi-component wildcard, so we have to evaluate all of them and then compare the results to determine which mapping is the best. though afaik we currently don't allow `@Parameters` that world map to (multi-component) wildcards, and also don't use them internally in Apodini. 


[^1]: this is something i added to match vapor's behavior: if there's an incoming HEAD request but no matching route exists we also look for a GET route that might match the request and use that if possible. what's your opinion on this? do you think this is desirable to have?

Reply:
Nice explanation. And I didn't know you could use footnotes this way in md, very nice, learned two things here 😅